File: python/vio_py/data_generator.py
from __future__ import annotations

import math

import numpy as np

from .vio_types import CameraPose, GeneratorConfig, Point3D, PointCloud, Trajectory


def generate_trajectory(config: GeneratorConfig | None = None) -> Trajectory:
    if config is None:
        config = GeneratorConfig()

    traj: Trajectory = []
    if config.num_poses <= 0:
        return traj

    denom = max(config.num_poses - 1, 1)
    for i in range(config.num_poses):
        t = float(i) / float(denom)
        angle = 2.0 * math.pi * config.helix_turns * t

        x = config.helix_radius * math.cos(angle)
        z = config.helix_radius * math.sin(angle)
        y = config.helix_height * t

        pos = np.array([x, y, z], dtype=np.float64)
        target = np.array([0.0, config.helix_height * 0.5, 0.0], dtype=np.float64)
        up = np.array([0.0, 1.0, 0.0], dtype=np.float64)

        forward = target - pos
        forward /= np.linalg.norm(forward) + 1e-12
        right = np.cross(forward, up)
        right /= np.linalg.norm(right) + 1e-12
        cam_up = np.cross(right, forward)
        cam_up /= np.linalg.norm(cam_up) + 1e-12

        T_wc = np.eye(4, dtype=np.float64)
        T_wc[:3, 0] = right
        T_wc[:3, 1] = cam_up
        T_wc[:3, 2] = -forward
        T_wc[:3, 3] = pos

        traj.append(CameraPose(timestamp=t, T_wc=T_wc))

    return traj


def generate_point_cloud(config: GeneratorConfig | None = None) -> PointCloud:
    if config is None:
        config = GeneratorConfig()

    cloud: PointCloud = []
    if config.num_points <= 0:
        return cloud

    rng = np.random.default_rng(seed=42)
    xs = rng.uniform(-config.cloud_extent, config.cloud_extent, size=config.num_points)
    ys = rng.uniform(0.0, config.cloud_height, size=config.num_points)
    zs = rng.uniform(-config.cloud_extent, config.cloud_extent, size=config.num_points)

    for x, y, z in zip(xs, ys, zs):
        t = float(y / config.cloud_height) if config.cloud_height > 0.0 else 0.0
        if t < 0.5:
            s = t * 2.0
            color = np.array([0.0, s, 1.0 - s], dtype=np.float64)
        else:
            s = (t - 0.5) * 2.0
            color = np.array([s, 1.0 - s, 0.0], dtype=np.float64)

        cloud.append(
            Point3D(
                position=np.array([x, y, z], dtype=np.float64),
                color=np.clip(color, 0.0, 1.0),
            )
        )

    return cloud

File: src/core/data_generator.cpp
#include "core/data_generator.h"

#include <Eigen/Geometry>

#include <cmath>
#include <random>

namespace vio {

Trajectory generateTrajectory(const GeneratorConfig& config) {
    Trajectory traj;
    traj.reserve(config.num_poses);

    for (int i = 0; i < config.num_poses; ++i) {
        double t = static_cast<double>(i) / (config.num_poses - 1);
        double angle = 2.0 * M_PI * config.helix_turns * t;

        double x = config.helix_radius * std::cos(angle);
        double y = config.helix_radius * std::sin(angle);
        double z = config.helix_height * t;

        Eigen::Vector3d pos(x, y, z);
        Eigen::Vector3d target(0.0, 0.0, config.helix_height * 0.5);
        Eigen::Vector3d up(0.0, 0.0, 1.0);

        Eigen::Vector3d forward = (target - pos).normalized();
        Eigen::Vector3d right = forward.cross(up).normalized();
        Eigen::Vector3d cam_up = right.cross(forward).normalized();

        Eigen::Matrix4d T_wc = Eigen::Matrix4d::Identity();
        T_wc.block<3, 1>(0, 0) = right;
        T_wc.block<3, 1>(0, 1) = cam_up;
        T_wc.block<3, 1>(0, 2) = -forward;
        T_wc.block<3, 1>(0, 3) = pos;

        CameraPose pose;
        pose.timestamp = t;
        pose.T_wc = T_wc;
        traj.push_back(pose);
    }

    return traj;
}

PointCloud generatePointCloud(const GeneratorConfig& config) {
    PointCloud cloud;
    cloud.reserve(config.num_points);

    std::mt19937 rng(42);
    std::uniform_real_distribution<double> dist_xz(-config.cloud_extent, config.cloud_extent);
    std::uniform_real_distribution<double> dist_y(0.0, config.cloud_height);

    for (int i = 0; i < config.num_points; ++i) {
        double x = dist_xz(rng);
        double y = dist_y(rng);
        double z = dist_xz(rng);

        float t = static_cast<float>(y / config.cloud_height);
        float r;
        float g;
        float b;
        if (t < 0.5f) {
            float s = t * 2.0f;
            r = 0.0f;
            g = s;
            b = 1.0f - s;
        } else {
            float s = (t - 0.5f) * 2.0f;
            r = s;
            g = 1.0f - s;
            b = 0.0f;
        }

        Point3D pt;
        pt.position = Eigen::Vector3d(x, y, z);
        pt.color = Eigen::Vector3f(r, g, b);
        cloud.push_back(pt);
    }

    return cloud;
}

} // namespace vio
